### ISSUE TITLE
Ordering of equations and variables in the runtime for Newton Diagnostics

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#10116]
@TeusvanderStelt's PR #9263 contains an implementation of the Newton convergence diagnostics of https://doi.org/10.1016/j.amc.2021.125991.

If you run NewtonDiagnostics.ThermoFluidSystem.M2 in the attached [NewtonDiagnostics.zip](https://github.com/OpenModelica/OpenModelica/files/10519632/NewtonDiagnostics.zip), you get the following results
```
   ****** Model name: NewtonDiagnostics.ThermoFluidSystem.M2
   ****** Initial                         : 1
   ****** Number of integer parameters    : 0
   ****** Number of discrete real params  : 0
   ****** Number of real parameters       : 13
   ******  1: id=1006, name=         A, value=  1.000000
   ******  2: id=1007, name=         Q, value=  4.000000
   ******  3: id=1008, name=       T_a, value=  6.000000
   ******  4: id=1009, name=       T_s, value=  0.000000
   ******  5: id=1010, name=         c, value=  1.000000
   ******  6: id=1011, name=       f_0, value=  1.000000
   ******  7: id=1012, name=   gamma_0, value=  1.000000
   ******  8: id=1013, name=   k_dummy, value=  0.000000
   ******  9: id=1014, name=       k_h, value=  0.200000
   ****** 10: id=1015, name=       k_p, value= 31.622777
   ****** 11: id=1016, name=        nu, value=  0.800000
   ****** 12: id=1017, name=       p_d, value=  1.000000
   ****** 13: id=1018, name=       p_s, value=  2.201000
   ****** Number of integer variables     : 0
   ****** Number of real variables        : 6
   ******  1: id=1000, name=       T_o, value=  3.996000
   ******  2: id=1001, name=         f, value=  0.999000
   ******  3: id=1002, name=     gamma, value=  0.999000
   ******  4: id=1003, name=       k_v, value=  0.999000
   ******  5: id=1004, name=       p_i, value=  2.198000
   ******  6: id=1005, name=       p_o, value=  1.998000

   Information about equations from non-linear pattern ....

               Total number of equations    = 6
               Number of independents       = 6
               Number of non-linear entries = 12

   Vector x0: all dependents ....

               x0[0] =   0.9990000000  (gamma)
               x0[1] =   0.9990000000  (k_v)
               x0[2] =   0.9990000000  (f)
               x0[3] =   2.1980000000  (p_i)
               x0[4] =   1.9980000000  (p_o)
               x0[5] =   3.9960000000  (T_o)

   Function values of all equations f(x0) .... 

               f^1 =  -0.0001999200
               f^2 =   0.0020020000
               f^3 =   0.0009995000
               f^4 =  -0.7330508076
               f^5 =   0.0003998000
               f^6 =   0.0079960000

   Function values of non-linear equations n(w0) .... 1

               n^1 = f^1 =  -0.0001999200
               n^2 = f^2 =   0.0020020000
               n^3 = f^3 =   0.0009995000
               n^4 = f^4 =  -0.7330508076
               n^5 = f^5 =   0.0003998000
               n^6 = f^6 =   0.0079960000

   Vector w0: non-linear dependents .... 

               w0[0] = x0[0] =   0.9990000000  (gamma)
               w0[1] = x0[1] =   0.9990000000  (k_v)
               w0[2] = x0[2] =   0.9990000000  (f)
               w0[3] = x0[3] =   2.1980000000  (p_i)
               w0[4] = x0[4] =   1.9980000000  (p_o)
               w0[5] = x0[5] =   3.9960000000  (T_o)

   Damping factor lambda =      1


  
   ========================================================  
   Final results   
   ========================================================

      alpha_i > 1e-02
      ------------------------
      alpha_4 =     0.224

      Gamma_ijk > 1e-02
      ------------------------
      Gamma_4_4_4 =     0.211

      |Sigma_jj| > 1e-02
      ------------------------
      Sigma_4_4 =     0.423

      Variables    Initial guess         |max(Gamma,Sigma)|
      ---------    ------------------    ------------------
      var_4         p_i =   2.198        0.423

      Equations    max(alpha,Gamma)
      ---------    ------------------
      eqn_4        0.224
```
These are the original equations in the paper
![immagine](https://user-images.githubusercontent.com/4233148/215138243-9d16b358-8d70-409d-8b6b-99de0e917310.png)
and these are the expected results from the paper
![immagine](https://user-images.githubusercontent.com/4233148/215138915-d1a4cde2-c783-4ba7-90e5-fe81fb397677.png) ![immagine](https://user-images.githubusercontent.com/4233148/215139247-0f46a530-502c-4d6f-8797-d62761b9331f.png)

The computed values are correct: for case number 2, there is only one alpha_4 = 0.224  > 1e-2, only one value |sigma_4_4| = 0.423 > 1e-2, and only one value of |Gamma_4_4_4| = 0.211 > 1e-2. 

Variable number 4 (for OMC's diagnostics) is identified to be p_i, and that is consistent with the result of the paper, where p_i is variable number six and the nonzero values are sigma_6_6 and Gamma_1_6_6. So far, so good.

The question is: what is equation number 4 in OMC (the one pointed by alpha_4 Gamma_4_4_4)? This should correspond to equation 1 in the paper, namely:
![immagine](https://user-images.githubusercontent.com/4233148/215142878-d6b9e830-8193-4bd0-86e6-9da5255b78d1.png)
Unfortunately the diagnostic code can only retrieve the number of the equation _in the strong component_, but not its actual content.

If I look at the declarative debugger output, I see
![immagine](https://user-images.githubusercontent.com/4233148/215143637-64f4f0ce-c3e6-4e5e-9109-70d70d501267.png)
unfortunately, the fourth equation in that list is not what I would expect.

The reason is that the code generation scrambles the order of the equations, for some reason. This is taken from `NewtonDiagnostics.ThermoFluidSystem.M2_06inz.c`:
```C
/*
equation index: 13
indexNonlinear: 0
type: NONLINEAR

vars: {gamma, k_v, f, p_i, p_o, T_o}
eqns: {6, 5, 4, 3, 2, 1}
*/
void NewtonDiagnostics_ThermoFluidSystem_M2_eqFunction_13(DATA *data, threadData_t *threadData)
```
so equation number 4 in the generated code (which is pointed at by the diagnostics) is the fourth element of {6, 5, 4, 3, 2, 1}, i.e., number 3, which is indeed the expected residual 
![immagine](https://user-images.githubusercontent.com/4233148/215142878-d6b9e830-8193-4bd0-86e6-9da5255b78d1.png)

To make things even worse, the order of the iteration variables reported by the declarative debugger is also different, namely:
![immagine](https://user-images.githubusercontent.com/4233148/215145190-92570814-2a58-4f1e-95ee-cfac41eac511.png)
I guess this may also be due to the fact that the debugger re-orders the variable in alphabetical order. This is convenient to find out a specific variable in a large list, but brings further confusion.

In any case, the runtime has access to the variable names, so identifying what are the guilty variables is not a problem. The problem is identifying the guilty equations.

For that, we need a mapping between the index of the unknowns of each strong component (this example only has one, but there could be many of them) and the corresponding equation index from the backend, so one can find it in the declarative debugger. Of course it doesn't make sense that one has to look for it in the C code.

@AnHeuermann, @kabdelhak, how can we retrieve such mapping in the runtime? Thanks!

**Comment by AnHeuermann** [comment]
> In any case, the runtime has access to the variable names, so identifying what are the guilty variables is not a problem. The problem is identifying the guilty equations.
>
>For that, we need a mapping between the index of the unknowns of each strong component (this example only has one, but there could be many of them) and the corresponding equation index from the backend, so one can find it in the declarative debugger. Of course it doesn't make sense that one has to look for it in the C code.

One could use the `<model name>_info.json` and parse it at runtime to find the defining equation number from the variable name. I guess we are already doing something like that to display error messages and so on, see [OMCompiler/SimulationRuntime/c/simulation/simulation_info_json.c](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/SimulationRuntime/c/simulation/simulation_info_json.c).

The relevant line in the M2_info.json would be:

```json
{"eqIndex":26,"section":"regular","tag":"system","display":"non-linear","unknowns":6,"defines":["p_i","gamma","f","T_o","k_v","p_o"],"equation":[[14,15,16,17,18,19],[20,21,22,23,24,25]]}
```

telling that one needs to evaluate `NewtonDiagnostics_ThermoFluidSystem_M2_eqFunction_26` to get `gamma` or call `void residualFunc26` to evaluate the residual. I'm not sure if the index 2 can be used to assume, that the second element of the resiual vector represents `gamma`, but in this case it does.

```C
/*
equation index: 26
indexNonlinear: 1
type: NONLINEAR

vars: {p_i, gamma, f, T_o, k_v, p_o}
eqns: {19, 18, 17, 16, 15, 14}
*/
void NewtonDiagnostics_ThermoFluidSystem_M2_eqFunction_26(DATA *data, threadData_t *threadData)
```

```C
void residualFunc26(RESIDUAL_USERDATA* userData, const double* xloc, double* res, const int* iflag)
```

**Comment by casella** [comment]
@AnHeuermann I'm afraid that won't work. All these variables and equations are inside a strong component, so the "definining equation" (i.e., the matching equation) of a given variable is arbitrary. We actually need the unique equation number from SimCode, which is what is displayed in the declarative debugger.

Anyway, @kabdelhak has a solution for that in the pipeline, he's just waiting for PR #9263 to be merged on master.

**Comment by AnHeuermann** [comment]
> We actually need the unique equation number from SimCode, which is what is displayed in the declarative debugger.

You could get the number of that equation, in this case `18`, but there is no C function to call. If you are lucky the variable is a inner variable, so we have a inner equation, but in general it is a iteration variable.

**Comment by casella** [comment]
@AnHeuermann I'm not sure I'm getting your point.

What we have here is a strong component, with N (iteration) variables and a N residual equations. The diagnostic algorithm will compute high indicators for "culprit nonlinear equations" to certain residuals, and I want to see who they are, to figure out if I can somehow tame their nonlinear behaviour, e.g. by means of homotopy.

If no tearing is performed (as in the examples of our paper, that we are currently testing) this is straighforward, as long as I get the correspondence between the SimCode equation number and the i-th element of the residual vector function of that strong component.

Of course in case tearing is applied, the residual equation is only part of the story, because some nonlinear behaviour could be buried in the inner equations, and we currently have no means to figure out how much each inner equation is influencing which residuals. I guess to get some additional intel, once we get a pointer to a specific residual equation, we could also trace back which inner equations are actually involved to compute it, and point them out as well, possibily with some quantitative information. That's one option for future work, but we haven't worked out the theory for that, yet. At the moment I am interested at getting the thing to work in the base case without tearing.

Are we on the same page?

**Comment by casella** [comment]
> Anyway, @kabdelhak has a solution for that in the pipeline, he's just waiting for PR #9263 to be merged on master.

@kabdelhak I'm afraid the process got stuck at some point, #9263 was merged a long time ago, do you still have that solution ready?

**Comment by kabdelhak** [comment]
I unearthed and revived an old branch of mine that seems to be aiming to fix this issue.

Let's see if it does the trick #12415

**Comment by kabdelhak** [comment]
it now reports additional information about the variable, i am not quite sure anymore what this was about to be honest.

I added a new structure to the nonlinear system data, which stores the simcode indices in order. i guess this was requested?
```
int *eqn_simcode_indices;
```

The report shows very small difference, but it seems to now show which variable belongs to this report:
https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-12415/1/tests

Is this what was requested?

**Comment by casella** [comment]
@kabdelhak, unfortunately this thread got stuck again for a long time. I hope we can now get through it. We started to work on this topic six years ago with @bernhardbachmann, it would be good to finally get through it.

I checked the report you linked https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-12415/1/tests, but I get an empty white page. What should it show? Can you get your PR through, so that @TeusvanderStelt can finalize it?

Thanks!

**Comment by casella** [comment]
@kabdelhak we're trying to resume this thread and get it though, we'll keep you posted.

**Comment by casella** [comment]
> I added a new structure to the nonlinear system data, which stores the simcode indices in order.
> 
> ```
> int *eqn_simcode_indices;

This looks good, but it shouldn't have changed the report, right?

> The report shows very small difference

I checked how you changed the code. I guess the information is OK, though I would need to check with the NewtonDiagnostics.ThermoFluidSystem.M2 example. The information is not displayed as I would have liked, but I think I can take care of that.

I'll add a commit to your PR #12415 ASAP.

**Comment by casella** [comment]
This should now be fixed by #12415
